### PR TITLE
Fix some broken opcodes due to JS number issues

### LINKED ISSUE #169: BLOCKHASH is broken
`getBlockHash` expects a `Buffer`:
- https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/stateManager.js#L213
- https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/fakeBlockChain.js#L6

But it always receives a Javascript number from the `BLOCKHASH` opcode: https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/opFns.js#L337

The reason it still works is due to ethereumjs-util in the background transform anything into `Buffer` before it converts them to number in `bufferToInt`.

### PR DESCRIPTION
**Pull request by axic**
Part of #167. Fixes #169.

### CODE REVIEW

**Comment by axic** [comment]
@cdetrio @jwasinger @holgerd77 dare to review this? :)

File: lib/opFns.js [hunk 1]
```diff
@@ -418,25 +421,37 @@ module.exports = {
...
+      dest = dest.toNumber()
 
-    runState.programCounter = dest
+      if (!jumpIsValid(runState, dest)) {
+        trap(ERROR.INVALID_JUMP + ' at ' + describeLocation(runState))
+      }
+
+      runState.programCounter = dest
+    }
   },
```
Comment (holgerd77) [inline lib/opFns.js]: Why not put the extra code length check here in the ``jumpIsValid`` logic?

File: lib/opFns.js [hunk 2]
```diff
@@ -418,25 +421,37 @@ module.exports = {
...
   JUMP: function (dest, runState) {
-    dest = utils.bufferToInt(dest)
+    dest = new BN(dest)
+
+    if (dest.gtn(runState.code.length)) {
+      trap(ERROR.INVALID_JUMP + ' at ' + describeLocation(runState))
+    }
+
+    dest = dest.toNumber()
```
Comment (holgerd77) [inline lib/opFns.js]: This looks good otherwise.

File: lib/opFns.js (hunk 1 above)
Comment (holgerd77) [inline lib/opFns.js]: Phew, this is so much better readable now. Looks good as well.

File: lib/fakeBlockChain.js [hunk 3]
```diff
@@ -3,6 +3,7 @@ const utils = require('ethereumjs-util')
 
 module.exports = {
   getBlock: function (n, cb) {
+    // FIXME: this will fail on block numbers >53 bits
```
Comment (axic) [inline lib/fakeBlockChain.js]: After upgrading ethereumjs-util it likely will throw an exception.

File: lib/opFns.js [hunk 4]
```diff
@@ -260,10 +258,16 @@ module.exports = {
...
 
-    loaded = loaded.length ? loaded : Buffer.from([0])
+    var loaded
+    if (pos > runState.callData.length) {
+      loaded = Buffer.from([0])
+    } else {
+      pos = pos.toNumber()
+      loaded = runState.callData.slice(pos, pos + 32)
+      loaded = loaded.length ? loaded : Buffer.from([0])
+    }
```
Comment (axic) [inline lib/opFns.js]: Good catch, it shouldn't work with `>` 😕

Javascript...

File: lib/opFns.js (hunk 1 above)
Comment (axic) [inline lib/opFns.js]: Hm, that is the edge case of having `JUMPI` as the last instruction in the code. Need to check the YP.

Comment (axic) [inline lib/opFns.js]: It seems that results in a `STOP` and not invalid jump. Will double check.

File: lib/opFns.js (hunk 4 above)
Comment (axic) [inline lib/opFns.js]: Fixed.

Comment (axic) [inline lib/opFns.js]: `console.log(new BN(0) > 42)` => `false`

Comment (holgerd77) [inline lib/opFns.js]: Hmmm, to just change this would have been the cleaner solution in my regard, but ok.

Comment (axic) [inline lib/opFns.js]: It was changed, just explained why it did "work" or more likely why the tests don't have a good coverage.

**Pull request merged**